### DISCUSSION TITLE
require-ing support code in a lua filter

### DISCUSSION THREAD

**Discussion by pragdave** [quarto-dev/quarto-cli#3188] [Q&A]
I'm writing a lua filter that I've split into a number of lua source files.

How can I require these files from the main filter script? (I've tried the `...` trick but that doesn't work... )

Cheers

Dave

**Comment by mcanouil** [comment]
I don't know what the `...` trick is supposed to be, but to import Lua script inside Lua script, Lua uses `require("lua-script-name-without-extension")`.
See <http://www.lua.org/pil/8.1.html>.

PS: could you change the category from "general" to "Q&A" (using the desktop version of the website, not mobile view or app)?

**Reply by dragonstyle** [reply]
As mcanouil said, using require is advised:

See more about require here: https://www.lua.org/pil/8.1.html

You can either use require to import the file and anything global will become accessible, or more preferable, the 'required' file can return a table of function, properties and so on which you can assign to a local variable- for example:

```lua
local importedStuff = require '<filename-without-extension>'
```

**Comment by pragdave** [comment]
I understand "require". My problem is how to specify a path that is relative to the file doing the requiring. For example, I might have

~~~
| index.qmd
| _quarto.yml
| _ extensions
    | filters
     _| my_filter.lua
      | lib_file1.lua
      | lib_file2.lua
~~~

In `my_filter.lua`, I want to be able to do something like:

``` lua
lib1 = require("./lib_file1")
lib2 = require("./lib_file2")
```

I can't, though, because if I run `quarto` from the top level, because the working directory it the top level.

OK, you say: make the require something like `require("_extensions/filter/..."). But then, how, can I distribute this filter?

TIA


Dave

**Reply by mcanouil** [reply]
A working reproducible example would make things a lot easier to understand thus easier to properly help you.

**Comment by pragdave** [comment]
I've worked around it. Thanks all.
